modules/localization/msf/common/util/frame_transform.cc

```diff
@@ -0,0 +1,474 @@
+ *
+ * Determines the central meridian for the given UTM zone.
+ *
+ * Inputs:
+ *     zone - An integer value designating the UTM zone, range [1,60].
+ *
+ * Returns:
+ *   The central meridian for the given UTM zone, in radians, or zero
+ *   if the UTM zone parameter is outside the range [1,60].
+ *   Range of the central meridian is the radian equivalent of [-177,+177].
+ *
+ */
+inline double UtmCentralMeridian(const int zone) {
+  return (-183.0 + (zone * 6.0)) * kSinsDegToRad;
+}
+
+/*
+ * FootpointLatitude
+ *
+ * Computes the footpoint latitude for use in converting transverse
+ * Mercator coordinates to ellipsoidal coordinates.
+ *
+ * Reference: Hoffmann-Wellenhof, B., Lichtenegger, H., and Collins, J.,
+ *   GPS: Theory and Practice, 3rd ed.  New York: Springer-Verlag Wien, 1994.
+ *
+ * Inputs:
+ *   y - The UTM northing coordinate, in meters.
+ *
+ * Returns:
+ *   The footpoint latitude, in radians.
+ *
+ */
+double FootpointLatitude(const double y) {
+  /* Precalculate n (Eq. 10.18) */
+  double n = (kSmA - kSmB) / (kSmA + kSmB);
+
+  /* Precalculate alpha (Eq. 10.22) */
+  /* (Same as alpha in Eq. 10.17) */
+  double alpha = ((kSmA + kSmB) / 2.0) * (1 + (pow(n, 2.0) / 4)
+                  + (pow(n, 4.0) / 64));
+
+  /* Precalculate yy (Eq. 10.23) */
+  double yy = y / alpha;
+
+  /* Precalculate beta (Eq. 10.22) */
+  double beta = (3.0 * n / 2.0)
+                 + (-27.0 * pow(n, 3.0) / 32.0)
+                 + (269.0 * pow(n, 5.0) / 512.0);
+
+  /* Precalculate gamma (Eq. 10.22) */
+  double gamma = (21.0 * pow(n, 2.0) / 16.0) + (-55.0 * pow(n, 4.0) / 32.0);
+
+  /* Precalculate delta (Eq. 10.22) */
+  double delta = (151.0 * pow(n, 3.0) / 96.0)
+                  + (-417.0 * pow(n, 5.0) / 128.0);
+
+  /* Precalculate epsilon (Eq. 10.22) */
+  double epsilon = (1097.0 * pow(n, 4.0) / 512.0);
+
+  /* Now calculate the sum of the series (Eq. 10.21) */
+  double result = yy + (beta * sin(2.0 * yy)) + (gamma * sin(4.0 * yy))
+           + (delta * sin(6.0 * yy)) + (epsilon * sin(8.0 * yy));
+
+  return result;
+}
+
+/*
+ * MaplatlonToXY
+ *
+ * Converts a latitude/longitude pair to x and y coordinates in the
+ * Transverse Mercator projection.  Note that Transverse Mercator is not
+ * the same as UTM; a scale factor is required to convert between them.
+ *
+ * Reference: Hoffmann-Wellenhof, B., Lichtenegger, H., and Collins, J.,
+ * GPS: Theory and Practice, 3rd ed.  New York: Springer-Verlag Wien, 1994.
+ *
+ * Inputs:
+ *    phi - Latitude of the point, in radians.
+ *    lambda - Longitude of the point, in radians.
+ *    lambda0 - Longitude of the central meridian to be used, in radians.
+ *
+ * Outputs:
+ *    xy - A 2-element array containing the x and y coordinates
+ *         of the computed point.
+ *
+ * Returns:
+ *    The function does not return a value.
+ *
+ */
+void MaplatlonToXY(const double phi, const double lambda,
+                   double lambda0, UTMCoor *xy) {
+  if (xy == nullptr) {
```

Comment:
I suggest using check in situations like this: CHECK_NOTNULL(xy); otherwise, please use a return status (bool MaplatlonToXy()), return false, and output some error msg.

Reply:
changed in all files